components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -3980,6 +3980,54 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         }
     }
 
+    protected _getOnboardingState: GitpodServer.OnboardingState | undefined;
+
+    async getOnboardingState(ctx: TraceContext): Promise<GitpodServer.OnboardingState> {
+        this.checkAndBlockUser("getOnboardingState");
+
+        if (this._getOnboardingState) {
+            return this._getOnboardingState;
+        }
+
+        // A shortcut for gitpod.io, but also for Dedicated PoC.
+        // This is controlled by feature flag (per Gitpod host.)
+        if (!this.enableDedicatedOnboardingFlow) {
+            return {
+                isCompleted: true,
+                hasAnyOrg: true,
+                hasAnyOrgWithInactiveSSO: false,
+            };
+        }
+
+        // Optimized check for completed setup
+        const hasAnyOrgWithActiveSSO = await this.teamDB.hasAnyOrgWithActiveSSO();
+        if (hasAnyOrgWithActiveSSO) {
+            // Let's cache the state at least per connection.
+            this._getOnboardingState = {
+                isCompleted: true,
+                hasAnyOrg: true,
+                hasAnyOrgWithInactiveSSO: false,
+            };
+            return this._getOnboardingState;
```

Comment:
Why? This would only really be encountered during the onboarding flow, and never after, no?

Reply:
This call will happen every time we load the dashboard for a logged in user, which is how it can determine if a setup flow needs to be presented.

Reply:
Yes. The motivation is to reduce load/delays by this calls, as the frontend would make them on _every_ initial load. There we can optimize as well, and memorize whether onboarding is completed.